src/Tasks/Microsoft.NET.Build.Tasks/RunCrossgen.cs

```diff
@@ -0,0 +1,387 @@
+        {
+            outputPDBImage = null;
+
+            string arguments = null;
+
+            if (_targetPlatform == OSPlatform.Windows)
+            {
+                if (_diasymreaderPath == null || !File.Exists(_diasymreaderPath))
+                    return false;
+
+                outputPDBImage = Path.ChangeExtension(file.ItemSpec, "ni.pdb");
+
+                arguments = $"/nologo " +
+                    $"/Platform_Assemblies_Paths \"{_platformAssembliesPath}{_pathSeparatorCharacter}{Path.GetDirectoryName(file.ItemSpec)}\" " +
+                    $"/DiasymreaderPath \"{_diasymreaderPath}\" " +
+                    $"/CreatePDB \"{Path.GetDirectoryName(file.ItemSpec)}\" " +
+                    $"\"{file.ItemSpec}\"";
+            }
+            else if(_targetPlatform == OSPlatform.Linux)
+            {
+                using (FileStream fs = new FileStream(file.ItemSpec, FileMode.Open, FileAccess.Read))
+                {
+                    PEReader pereader = new PEReader(fs);
+                    MetadataReader mdReader = pereader.GetMetadataReader();
+                    Guid mvid = mdReader.GetGuid(mdReader.GetModuleDefinition().Mvid);
+
+                    outputPDBImage = Path.ChangeExtension(file.ItemSpec, "ni.{" + mvid + "}.map");
+                }
+
+                arguments = $"/nologo " +
+                    $"/Platform_Assemblies_Paths \"{_platformAssembliesPath}{_pathSeparatorCharacter}{Path.GetDirectoryName(file.ItemSpec)}\" " +
+                    $"/CreatePerfMap \"{Path.GetDirectoryName(file.ItemSpec)}\" " +
+                    $"\"{file.ItemSpec}\"";
+            }
+
+            return arguments == null ? false : RunCrossgenProcessAndLog(arguments);
+        }
+
+        bool RunCrossgenProcessAndLog(string arguments)
+        {
+            Process crossgenProcess = new Process
+            {
+                StartInfo = new ProcessStartInfo
+                {
+                    CreateNoWindow = true,
+                    Arguments = arguments,
+                    FileName = _crossgenPath,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true
+                }
+            };
+
+            crossgenProcess.Start();
+            string stdout = crossgenProcess.StandardOutput.ReadToEnd();
+            string stderr = crossgenProcess.StandardError.ReadToEnd();
+            crossgenProcess.WaitForExit();
+
+            if (crossgenProcess.ExitCode != 0)
+            {
+                Log.LogError(Strings.ReadyToRunCompilerOutput, stderr.Trim());
+                return false;
+            }
+
+            Log.LogMessage(MessageImportance.Normal, Strings.ReadyToRunCompilerOutput, stdout.Trim());
+
+            return true;
+        }
+
+        bool ExtractTargetPlatformAndArchitectureFromRuntimeIdentifier()
+        {
+            string targetPlatform = RuntimeIdentifier.Split(new[] { '-' }, StringSplitOptions.RemoveEmptyEntries)[0].ToLower();
```

Comment:
The logic in this function seems like it could be useful to other tasks too (if not already implemented somewhere else that I couldn't find). Where would you recommend we put this?

Reply:
Are there variables in the SDK that we can read instead of doing this?

Reply:
Leaving it as is for now, as we discussed